josh-core/src/history.rs

```diff
@@ -196,49 +196,70 @@ pub struct RewriteData<'a> {
     pub message: Option<String>,
 }
 
-// takes everything from base except it's tree and replaces it with the tree
+// takes everything from base except its tree and replaces it with the tree
 // given
 pub fn rewrite_commit(
     repo: &git2::Repository,
-    base: &git2::Commit,
+    base: &git2::Commit, // original commit that we are modifying
     parents: &[&git2::Commit],
     rewrite_data: RewriteData,
     unsign: bool,
 ) -> JoshResult<git2::Oid> {
-    let message = rewrite_data
-        .message
-        .unwrap_or(base.message_raw().unwrap_or("no message").to_string());
-    let tree = &rewrite_data.tree;
-
-    let a = base.author();
-    let new_a = if let Some((author, email)) = rewrite_data.author {
-        git2::Signature::new(&author, &email, &a.when())?
-    } else {
-        a
-    };
+    use bstr::ByteSlice;
+    use gix_object::{CommitRef, WriteTo};
 
-    let c = base.committer();
-    let new_c = if let Some((committer, email)) = rewrite_data.committer {
-        git2::Signature::new(&committer, &email, &c.when())?
-    } else {
-        c
-    };
+    let odb = repo.odb()?;
+    let odb_commit = odb.read(base.id())?;
+    assert!(odb_commit.kind() == git2::ObjectType::Commit);
 
-    let b = repo.commit_create_buffer(&new_a, &new_c, &message, tree, parents)?;
+    // gix_object uses BStr for Oids, but in hex representation, not raw bytes. Because of lifetimes we have to
+    // pre-prepare this before creating the CommitRef
+    let tree_id = format!("{}", rewrite_data.tree.id());
+    let parent_ids = parents
+        .iter()
+        .map(|x| format!("{}", x.id()))
+        .collect::<Vec<_>>();
+
+    let mut commit = CommitRef::from_bytes(odb_commit.data())?;
+
+    commit.tree = tree_id.as_bytes().into();
+
+    commit.parents.clear();
+    commit
+        .parents
+        .extend(parent_ids.iter().map(|x| x.as_bytes().into()));
+
+    if let Some(ref msg) = rewrite_data.message {
+        commit.message = msg.as_bytes().into();
+    }
+
+    if let Some((ref author, ref email)) = rewrite_data.author {
+        commit.author.name = author.as_bytes().into();
+        commit.author.email = email.as_bytes().into();
+    }
+
+    if let Some((ref author, ref email)) = rewrite_data.committer {
+        commit.committer.name = author.as_bytes().into();
+        commit.committer.email = email.as_bytes().into();
+    }
+
+    commit
+        .extra_headers
+        .retain(|(k, _)| k.as_bytes() != "gpgsig".as_bytes());
+
+    let mut b = vec![];
+    commit.write_to(&mut b)?;
 
     if let (false, Ok((sig, _))) = (unsign, repo.extract_signature(&base.id(), None)) {
```

Comment:
Would't it be possible to combine the the handling of the signature with the remaining headers. That is,
` commit
         .extra_headers
         .retain(|(k, _)| (k.as_bytes() != "gpgsig".as_bytes()) || !unsign);`

and getting rid of the extra `commit_signed` below?

I think it might even necessary to do so depending on how headers ordering is handled/specified (or rather unspecified)

Reply:
I'll give that a try and see if the tests pass!